// [python quickstart](https://grpc.io/docs/quickstart/python.html#run-a-grpc-application)
// python -m grpc_tools.protoc --python_out=. --grpc_python_out=. -I. correspondence.proto

// backend.proto
syntax = "proto3";

service Backend {
    // the demo for test
    rpc SayHello(HelloRequest) returns (HelloResponse) {}
    rpc SayHelloAgain(HelloRequest) returns (HelloResponse) {}

    // main interface
    // ->interface on database interaction
    // ->->on note table
    rpc GetAllNotes(RequestStruct) returns (ResponseStruct) {} // 教师学生共用，通过角色类型划分权限
    rpc GetTheNote(RequestStruct) returns (ResponseStruct) {} // 教师学生共用，通过id获取
    rpc InsertANote(RequestStruct) returns (ResponseStruct) {} // 教师专用
    rpc ModifyTheNote(RequestStruct) returns (ResponseStruct) {} // 教师专用
    // ->->


}


// to avoid some trouble, all data will be serialized before transmission and anti-serialized after transmission
// will be serialized and anti-serialized by module `pickle`
// data structure for test
message HelloRequest {
    bytes para = 1;
}

message HelloResponse {
    bytes result = 1;
}

// main data structure
message RequestStruct {
    bytes para = 1;
}

message ResponseStruct {
    bytes result = 1;
}